cpp/src/reductions/histogram.cu

```diff
@@ -0,0 +1,270 @@
+                                                                                   d_hasher,
+                                                                                   d_equal,
+                                                                                   d_output},
+      d_partial_output{d_partial_output}
+  {
+  }
+
+  // Count the number of rows in each group of rows that are compared equal.
+  __device__ void operator()(size_type const idx) const
+  {
+    auto const increment = d_partial_output ? d_partial_output[idx] : CountType{1};
+    auto const count =
+      cuda::atomic_ref<CountType, cuda::thread_scope_device>(*this->get_output_ptr(idx));
+    count.fetch_add(increment, cuda::std::memory_order_relaxed);
+  }
+};
+
+/**
+ * @brief The builder to construct an instance of `reduce_fn` functor.
+ */
+template <typename CountType>
+struct reduce_func_builder {
+  CountType const* const d_partial_output;
+
+  reduce_func_builder(CountType const* const d_partial_output) : d_partial_output{d_partial_output}
+  {
+  }
+
+  template <typename MapView, typename KeyHasher, typename KeyEqual>
+  auto build(MapView const& d_map,
+             KeyHasher const& d_hasher,
+             KeyEqual const& d_equal,
+             CountType* const d_output)
+  {
+    return reduce_fn<MapView, KeyHasher, KeyEqual, CountType>{
+      d_map, d_hasher, d_equal, d_output, d_partial_output};
+  }
+};
+
+/**
+ * @brief Specialized functor to check for non-zero of the second component of the input.
+ */
+struct is_none_zero {
+  template <typename Pair>
+  __device__ bool operator()(Pair const input) const
+  {
+    return thrust::get<1>(input) != 0;
+  }
+};
+
+/**
+ * @brief Building a histogram by gathering distinct rows from the input table and their
+ * corresponding distinct counts.
+ *
+ * @param input The input table
+ * @param distinct_indices Indices of the distinct rows
+ * @param distinct_counts Distinct counts corresponding to the distinct rows
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate the returned object's device memory
+ * @return A list_scalar storing the output histogram
+ */
+auto gather_histogram(table_view const& input,
+                      device_span<size_type const> distinct_indices,
+                      std::unique_ptr<column>&& distinct_counts,
+                      rmm::cuda_stream_view stream,
+                      rmm::mr::device_memory_resource* mr)
+{
+  auto distinct_rows =
+    std::move(cudf::detail::gather(input,
+                                   distinct_indices,
+                                   out_of_bounds_policy::DONT_CHECK,
+                                   cudf::detail::negative_index_policy::NOT_ALLOWED,
+                                   stream,
+                                   mr)
+                ->release()
+                .front());
```

Comment:
Should this be broken into multiple expressions? clang-format could not handle it well :D
Maybe a lambda if you don't want intermediate variables visible in the rest of the function